src/toil/job.py

```diff
@@ -3758,6 +3776,374 @@ def getUserScript(self):
+    ):
+        """
+        Setup importing files on a worker.
+        :param filenames: List of file URIs to import
+        :param disk_size: Designated disk space the worker can use when importing. Disregarded if stream is enabled.
+        :param stream: Whether to stream a file import or not. We don't have machinery to ensure
+        streaming, so if true, assume streaming works and don't give the worker a lot of disk space to work with.
+        If streaming fails, the worker will run out of resources and allocate a child job to handle the import with enough disk space.
+        :param kwargs: args for the superclass
+        """
+        if stream:
+            super().__init__(local=False, **kwargs)
+        else:
+            super().__init__(local=False, disk=disk_size, **kwargs)
+        self.filenames = filenames
+        self.disk_size = disk_size
+        self.stream = stream
+
+    @staticmethod
+    def import_files(
+        files: List[str], file_source: "AbstractJobStore"
+    ) -> Dict[str, FileID]:
+        """
+        Import a list of files into the jobstore. Returns a mapping of the filename to the associated FileIDs
+
+        When stream is true but the import is not streamable, the worker will run out of
+        disk space and run a new import job with enough disk space instead.
+        :param files: list of files to import
+        :param file_source: AbstractJobStore
+        :return: Dictionary mapping filenames to associated jobstore FileID
+        """
+        # todo: make the import ensure streaming is done instead of relying on running out of disk space
+        path_to_fileid = {}
+
+        @memoize
+        def import_filename(filename: str) -> Optional[FileID]:
+            return file_source.import_file(filename, symlink=True)
+
+        for file in files:
+            imported = import_filename(file)
+            if imported is not None:
+                path_to_fileid[file] = imported
+        return path_to_fileid
+
+    def run(self, file_store: AbstractFileStore) -> Promised[Dict[str, FileID]]:
+        """
+        Import the workflow inputs and then create and run the workflow.
+        :return: Promise of workflow outputs
+        """
+        try:
+            return self.import_files(self.filenames, file_store.jobStore)
+        except OSError as e:
+            # If the worker crashes due to running out of disk space and was not trying to
+            # stream the file import, then try a new import job without streaming by actually giving
+            # the worker enough disk space
+            # OSError 28 is no space left on device
+            if e.errno == 28 and self.stream is True:
+                non_streaming_import = WorkerImportJob(
+                    self.filenames, self.disk_size, stream=False
+                )
+                self.addChild(non_streaming_import)
+                return non_streaming_import.rv()
+            else:
+                raise
```

Comment:
I don't think this is a good approach; we don't always sandbox jobs to keep them within their requested disk space, so if one job goes over its disk request it can make a *different* job fail due to not having enough space for *its* temporary files.

This might honestly be better if we *didn't* handle the case where the attempted streaming import went over its disk space limit, and just made it the user's problem to set a disk space limit big enough if they have any imports that can't actually stream?

Or we could implement a flag on the job store import method to only allow streaming and fail if streaming is not possible, and catch *that* and do the fallback.

Also, if we hit the disk space limit after importing several files already, when do we delete those imported copies? It looks like we will leave them behind in the job store and then re-import the same files non-streaming with more disk space.

Reply:
I think I'll leave it up to the user to deal with if this is too complicated to support.